### PR TITLE
Add slash command to set tokenizer

### PR DESCRIPTION
**Pull request by eugencowie**
This adds a `/tokenizer` slash command which can be used to set the tokenizer to any supported value (including 'best_match'). This is useful if you use quick replies for changing api/model/presets and want to set the tokenizer as well.

Implements #2575

image

First contribution to ST so not very familiar with the codebase, please let me know if this could have been implemented in a better way.

## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

### CODE REVIEW

File: public/scripts/tokenizers.js [hunk 1]
```diff
@@ -147,6 +147,24 @@ async function resetTokenCache() {
     }
 }
 
+/**
+ * Maps tokenizer IDs to their names.
+ * @example { 0: 'none', 1: 'gpt2', ... }
+ */
+export const TOKENIZER_NAME_MAP = Object.fromEntries(
```
Comment (Wolfsblvt) [inline public/scripts/tokenizers.js]: Isn't `tokenizers` aready an exported enum that contains both friendly names and the id/value that is needed?
I mean doesn't hurt, so it's okay, but if it's not much more verbose to write simple `.map` inside the slash command enum and callback, we don't need to make another list here.

Generally saying, as much as I don't like going directly for UI elements in backend code.. but as we already need to set and trigger the event on the `#tokenizer` anyway, you *could* even just query the option list from that. It contains both the exact ids you need, and also the friendly names, already in a formatted way that works nice to present in the slash command.

Reason why I suggest this: The `tokenizers` map does contain more tokenizers than the ones that are available in the UI. Likely on purpose, though @Cohee1207 would have to say that.
The option 2: **OPENAI** is missing.

So if you select that one from your slash command, it'll set it to null and that leads to errors.

**Review by Wolfsblvt — CHANGES_REQUESTED** [review]
Looks very solid! Thanks for the contribution.

Small issue with the inconsistent tokenizer list that I pointed out, that we need to solve first.

File: public/scripts/tokenizers.js (hunk 1 above)
Comment (Cohee1207) [reply public/scripts/tokenizers.js]: OPENAI tokenizer only exists in code to be used as a route for Chat Completion API. I agree that using UI elements might be beneficial for two reasons. 1) Filter out tokenizers that only exist as auxiliary values. 2) Get proper display names for the enum provider.

Comment (eugencowie) [reply public/scripts/tokenizers.js]: Thanks for the feedback, I've updated it to get the available tokenizers from the UI.

**Review by Wolfsblvt — APPROVED** [review]
Just realized that this might not be the *best* idea to accept the real friendly names as the key in scripts. Should've thought that through...

I made some slight adjustments so you don't have to do it all over again.
It'll now take the internal key (enum key) as the actual return value, and also prefer this as the input to set - but thanks to Fuse you can also search the friendly name.

I think this is pretty enough.

image

I also took the liberty to change some JS types, so the new function and structure is consistent to the existing `getFriendlyTokenizerName`.

Approved and tested now. Feel free to check if this works for you.

**Review by Cohee1207 — APPROVED** [review]
Moved the command from script.js to a more appropriate place, I think.

**Comment by Wolfsblvt** [comment]
> Moved the command from script.js to a more appropriate place, I think.

Well, the *most* appropriate place would be just inside the `tokenizers.js` file itself, wouldn't it?
Like tags, world info etc, functionality that has their separate files could have the slash commands registered there. Saves imports and structures it logically.

**Comment by Cohee1207** [comment]
Maybe. But anything is better than `script.js`

**Pull request merged**